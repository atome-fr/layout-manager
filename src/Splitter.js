import React, {Component} from 'react';

/**
 * Class used to manage a splitter
 * @class
 */
class Splitter extends Component {

    /**
     * Override constructor
     * @param props
     */
    constructor(props) {
        super(props);

        this.handleMouseOrTouchStart = this.handleMouseOrTouchStart.bind(this);
        this.handleMouseOrTouchMove = this.handleMouseOrTouchMove.bind(this);
        this.handleMouseOrTouchEnd = this.handleMouseOrTouchEnd.bind(this);
    }

    /**
     * Handler when the user start to move the splitter
     */
    handleMouseOrTouchStart(ev) {
        if ((ev.clientX && ev.clientY) || (ev.touches && ev.touches.length > 0)) {
            this.mousePos = {x: ev.clientX || ev.touches[0].clientX, y: ev.clientY || ev.touches[0].clientY};
        }
        document.addEventListener('mousemove', this.handleMouseOrTouchMove);
        document.addEventListener('touchmove', this.handleMouseOrTouchMove);
        document.addEventListener('mouseup', this.handleMouseOrTouchEnd);
        document.addEventListener('touchend', this.handleMouseOrTouchEnd);
    }

    /**
     * Handler when the user move the splitter
     * @param ev
     */
    handleMouseOrTouchMove(ev) {
        ev.preventDefault();

        let newMousePos = undefined;
        if ((ev.clientX && ev.clientY) || (ev.touches && ev.touches.length > 0)) {
            newMousePos = {x: ev.clientX || ev.touches[0].clientX, y: ev.clientY || ev.touches[0].clientY};
        }

        if(newMousePos) {
            const offset = {x: newMousePos.x - this.mousePos.x, y: newMousePos.y - this.mousePos.y};
            this.mousePos = newMousePos;
            this.props.onChange(offset, this.el);
        }
    }

    /**
     * Handler when the user end to move the splitter
     */
    handleMouseOrTouchEnd() {
        document.removeEventListener('mousemove', this.handleMouseOrTouchMove);
        document.removeEventListener('touchmove', this.handleMouseOrTouchMove);
        document.removeEventListener('mouseup', this.handleMouseOrTouchEnd);
        document.removeEventListener('touchend', this.handleMouseOrTouchEnd);
    }

    /**
     * Method for render the components
     * @returns {*}
     */
    render() {
        return (
            <div ref={(el) => this.el = el}
                 className={"splitter " + this.props.type}
                 onTouchStart={this.handleMouseOrTouchStart}
                 onMouseDown={this.handleMouseOrTouchStart}>
                <div className="handle">
                </div>
            </div>
        );
    }
}

export default Splitter;