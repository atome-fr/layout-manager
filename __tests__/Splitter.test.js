import React from 'react';
import Splitter from '../src/Splitter';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

describe('Splitter Component', () => {

    describe('UI Render Test', () => {

        it('should render', () => {
            const wrapper = shallow(<Splitter type=""/>);
            const tree = toJson(wrapper);

            expect(tree).toMatchSnapshot();
        });

        it('should call the correct callback when a touch start or a mouse down is trigger on the splitter', () => {
            const oldVal = Splitter.prototype.handleMouseOrTouchStart;
            Splitter.prototype.handleMouseOrTouchStart = jest.fn();
            const wrapper = shallow(<Splitter/>);

            expect(Splitter.prototype.handleMouseOrTouchStart).toHaveBeenCalledTimes(0);

            wrapper.simulate('touchstart');

            expect(Splitter.prototype.handleMouseOrTouchStart).toHaveBeenCalledTimes(1);

            wrapper.simulate('mousedown');

            expect(Splitter.prototype.handleMouseOrTouchStart).toHaveBeenCalledTimes(2);

            Splitter.prototype.handleMouseOrTouchStart = oldVal;
        });

    });

    describe('Component Method Test', () => {

        it('should add some listeners on the document and store the position of the click', () => {
            const oldVal = document.addEventListener;
            document.addEventListener = jest.fn();

            const wrapper = shallow(<Splitter />);

            const ev = {
                clientX: 10,
                clientY: 10
            };

            wrapper.instance().handleMouseOrTouchStart(ev);

            expect(wrapper.instance().mousePos).toEqual({x: 10, y: 10});
            expect(document.addEventListener).toHaveBeenCalledTimes(4);
            expect(document.addEventListener.mock.calls[0]).toEqual(["mousemove", expect.any(Function)]);
            expect(document.addEventListener.mock.calls[1]).toEqual(["touchmove", expect.any(Function)]);
            expect(document.addEventListener.mock.calls[2]).toEqual(["mouseup", expect.any(Function)]);
            expect(document.addEventListener.mock.calls[3]).toEqual(["touchend", expect.any(Function)]);

            document.addEventListener = oldVal;
        });

        it('should handle the move position and call a callback method', () => {
            const handleOnChange = jest.fn();

            const wrapper = shallow(<Splitter onChange={handleOnChange}/>);

            wrapper.instance().mousePos = {x: 10, y: 10};

            const ev = {
                clientX: 15,
                clientY: 15,
                preventDefault: jest.fn()
            };

            wrapper.instance().handleMouseOrTouchMove(ev);

            expect(ev.preventDefault).toHaveBeenCalledTimes(1);
            expect(handleOnChange).toHaveBeenCalledTimes(1);
        });

        it('should remove some listeners on the document', () => {
            const oldVal = document.removeEventListener;
            document.removeEventListener = jest.fn();

            const wrapper = shallow(<Splitter />);

            wrapper.instance().handleMouseOrTouchEnd();

            expect(document.removeEventListener).toHaveBeenCalledTimes(4);
            expect(document.removeEventListener.mock.calls[0]).toEqual(["mousemove", expect.any(Function)]);
            expect(document.removeEventListener.mock.calls[1]).toEqual(["touchmove", expect.any(Function)]);
            expect(document.removeEventListener.mock.calls[2]).toEqual(["mouseup", expect.any(Function)]);
            expect(document.removeEventListener.mock.calls[3]).toEqual(["touchend", expect.any(Function)]);

            document.removeEventListener = oldVal;
        });

    });

});