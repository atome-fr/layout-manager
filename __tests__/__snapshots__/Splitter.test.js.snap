// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Splitter Component UI Render Test should render 1`] = `
<div
  className="splitter "
  onMouseDown={[Function]}
  onTouchStart={[Function]}
>
  <div
    className="handle"
  />
</div>
`;
